### PR TITLE
Feature/update contributing

### LINKED ISSUE #176: Community Member: Update Contributing Guidelines
Our contributing guidelines have fallen behind our actual process. Contributing Guidelines should be updated to reflect the current state of the world

```gherkin
Feature: Update Contributing Guidelines
  As a community member
  I want to know how to contribute to the Punchcard codebases
  So that I can participate effectively in the Punchcard community

  Scenario: Versioning and Changelog
    Given that I want to create a new release
     When I read the contributing guidelines for creating a new release
     Then I should understand how new versions are released
      And I should understand how changelogs are created

  Scenario: New Issue
    Given that I want to create a new issue or pull request
     When I read the contributing guidelines for creating a new issue
     Then I should see latest templates for creating issues
      And have links to do so using the latest templates available

  Scenario: Commit Message
    Given that I want to create a new commit
     When I read the contributing guidelines for writing commits
     Then I should see an template version of a commit
      And I should see an example commit

  Scenario: Pull Request Etiquette
    Given that I want to create a new pull request
     When I read the contributing guidelines for creating a new pull request
     Then I should understand the expectations, dos, and don'ts when creating a pull request
      And I should understand how the team defines a pull request as ready to merge
      And I should understand what the team considers out-of-scope for a pull request

  Scenario: Code of Conduct
    Given that I want to contribute to Punchcard
     When I read the contributing guidelines for contributing
     Then I should understand how the Code of Conduct is applied to my contributions

Feature: Add Supporting Files
  As a community member
  I want to have supporting files for the contributing guidelines
  So that I can easily reference those items

  Scenario: Code of Conduct
    Given that I want to participate in the Punchcard community
     When I want to understand what is acceptable participation standards 
      And I want to understand how those participation standards are enforced
     Then I should be able to read and understand the Punchcard Code of Conduct
      And I should know how I can report behavior contrary to the Code of Conduct
      And I should know what steps are taken to remove offending actions or actors from the community

Feature: Clean Crufty Files
  As a community member
  I want to see only relevant, up-to-date participation files
  So that I'm not confused about what is and is not up to date

  Scenario: Changelog
    Given that I to see the project's Changelog
     When I read the Changelog 
     Then I should understand that the manual Changelog has been deprecated
      And I should be directed to the project's releases
```

---

**Update Contributing Guidelines**
- [x] Versioning and Changelog (https://github.com/Snugug/punchcard/commit/a70a57ba917c3c2d55c016fe81c3562ebfae87da)
- [x] New Issue (https://github.com/Snugug/punchcard/commit/22573d6efeedbe07b2df5a1b522813571e8b346e)
- [x] Commit Messages (https://github.com/Snugug/punchcard/commit/13ac7d09ac37671aff6dc97358b43d8010feadc5)
- [x] Pull Request Etiquette (https://github.com/Snugug/punchcard/commit/ec9d6bda0fe100d9409c6d34e853833582e92f28)
- [x] Code of Conduct (https://github.com/Snugug/punchcard/commit/bd8eaf5a0b0ed1c79e4b14220e735c69fee7c3bf)

**Add Supporting Files**
- [x] Code of Conduct (https://github.com/Snugug/punchcard/commit/bd8eaf5a0b0ed1c79e4b14220e735c69fee7c3bf)

**Clean Crufty Files**
- [x] Changelog (https://github.com/Snugug/punchcard/commit/e5cad5da7f4b0a9089e98571cccef0540f66dbdc)

### PR DESCRIPTION
**Pull request by Snugug**
Update Contributing Guidelines for current issue templates, PR etiquette, and code of conduct

---
Resolves #176 

`DCO 1.1 Signed-off-by: Sam Richard <sam@snug.ug>`

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -19,18 +21,36 @@ Punchcard is a headless CMS designed with content strategy at its heart. With a
...
-* Include a summarized list of changes, additions, and deletions in the body of the pull request in the same style as [commit messages](#git-commit-messages).
-* Follow our JavaScript and CSS styleguides. We have linters set up to catch most of it.
-* Ensure that [EditorConfig](http://editorconfig.org/) installed in the editor used to work on the site and that it is functioning properly.
-* Do not squash or rebase commits when submitting a Pull Request. It makes it much harder to follow work and make incremental changes.
+* Do not squash or rebase commits when submitting a Pull Request
 * Ensure no Emoji tags are used in the title of the Pull Request
 
+## Pull Request Guidelines
+
+In order to expedite the process of reviewing and merging pull requests, we request the following guidelines are followed.
```
Comment (Snugug) [inline CONTRIBUTING.md]: Following are followed, or something like `following guidelines are adhered to`?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -19,18 +21,36 @@ Punchcard is a headless CMS designed with content strategy at its heart. With a
...
+* For consistency, follow the conventions that the team has in place that aren't necessarily covered by our automated test suite. For instance, the team may be using `map` instead of `forEach`, or partial conventions in Sass.
+* When writing browser code, ensure it works cross-browser in all Evergreen (auto-updating, not tied to a operating system version, not referred to by version number) browsers, as well as the latest version of Safari on iOS devices.
+* When writing browser code, we adhere to the principles of [progressive enhancement](http://alistapart.com/article/understandingprogressiveenhancement) and expect new code to follow the same principles.
+* Solve the issue in as few lines of code, with as few external dependencies, as possible. The most maintainable lines of code are the ones that don't exist, so keeping PRs as small, clear, and concise as possible will aid in overall project maintainability, stability, and make it easier for us to do code review.
+* Commit tests separately from implementation code.
+* Commit early and often. Small, atomic commits help us understand the thought process that went in to creating a pull request and make it easier for us to review.
```
Comment (Snugug) [inline CONTRIBUTING.md]: Should we reinforce the no squashing or rebasing of commits here?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -19,18 +21,36 @@ Punchcard is a headless CMS designed with content strategy at its heart. With a
...
+* Do not [refactor](http://blog.codeclimate.com/blog/2014/01/09/when-is-it-time-to-refactor/) existing code unless it is _absolutely necessary_ to resolve the issue being worked on. If there is an opportunity to refactor, please file a separate issue to discuss and implement instead.
+* Only include code that resolves the scenarios in the issue being worked on. While appreciated, work that goes above-and-beyond the scenarios outlined in an issue is out-of-scope and we will not be able to accept it. Please either work with the Punchcard team to get scenarios written for the issue being worked on, or split up the work across multiple issues and pull requests.
```
Comment (Snugug) [inline CONTRIBUTING.md]: Should `scenarios` link to something to describe them?

File: CONTRIBUTING.md (hunk 1 above)
Comment (scottnath) [inline CONTRIBUTING.md]: `In order to expedite the process of reviewing and merging pull requests, please adhere to all of the following guidelines.`
or
`In order to expedite the process of reviewing and merging pull requests, you must adhere to _all_ of the following guidelines.`

File: CONTRIBUTING.md (hunk 2 above)
Comment (scottnath) [inline CONTRIBUTING.md]: yes, but briefly!

File: CONTRIBUTING.md (hunk 1 above)
Comment (Snugug) [inline CONTRIBUTING.md]: Generally, `you` should be avoided in professional writing like this. I actually have a todo to go back and clean up the language of the contributing guidelines to remove the `you`s that exist

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -19,18 +21,36 @@ Punchcard is a headless CMS designed with content strategy at its heart. With a
...
+
+In order to expedite the process of reviewing and merging pull requests, we request the following guidelines are followed.
+
+### Tips for Submitting Great PRs
+
+Beyond just writing the code needed to resolve an issue, there are a few best practices that should be followed to submit a great PR:
+
+* We have [EditorConfig](http://editorconfig.org/) files set up for Punchcard. Having EditorConfig installed and used will help ensure that editors are configured to follow our conventions.
+* We have JavaScript and Sass linting set up (where appropriate). Before committing code, make sure that linting has been run locally and passes by running `npm run lint`
+* We have automated testing set up. It can be run locally by by running `npm run test`. This will run linting, our test suite, and show code coverage.
```
Comment (scottnath) [inline CONTRIBUTING.md]: this is `npm test`

Comment (scottnath) [inline CONTRIBUTING.md]: also...should there be a message here that PR will automatically fail travis if they fail these tests?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -19,18 +21,36 @@ Punchcard is a headless CMS designed with content strategy at its heart. With a
...
+* For consistency, follow the conventions that the team has in place that aren't necessarily covered by our automated test suite. For instance, the team may be using `map` instead of `forEach`, or partial conventions in Sass.
```
Comment (scottnath) [inline CONTRIBUTING.md]: I don't love this caveat...but mostly the distinction for using `map` vs `forEach`. If one is going over an array, but not planning to return anything back out...`forEach` can be totally appropriate.

File: CONTRIBUTING.md (hunk 3 above)
Comment (scottnath) [inline CONTRIBUTING.md]: Yes...is that the `writing issues` wiki entry? I think that needs a little updating.

File: CONTRIBUTING.md (hunk 5 above)
Comment (Snugug) [inline CONTRIBUTING.md]: Happy to either clarify that particular statement, or use a different example. What do you suggest?

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -72,71 +92,79 @@ By making a contribution to this project, I certify that:
 
 ### Git Commit Messages
 
+Commit message should follow the template `:<emoji>: <subject>` with `<emoji>` being the name of the relevant [emoji](#emoji-cheatsheet) describing the changes (without wrapping `::`) and `subject` being the description of changes. Commit messages may have multiple emoji. Commit messages should follow the following guidelines:
```
Comment (scottnath) [inline CONTRIBUTING.md]: This is totally fine....but the actual emoji-does-xyz needs to be fleshed out to match what's happening in our semvar module. A couple times I've had to dig in to the code on semvar to figure out which emoji would trigger what. We should have the specific emoji that do XYZ in the readme of semvar (or somewhere more apparent...or maybe in this doc?).

File: CONTRIBUTING.md (hunk 4 above)
Comment (Snugug) [inline CONTRIBUTING.md]: Further down this list I say that all required automated checks must pass, and I guess I imply this, but I can make it explicit if you'd prefer

Comment (Snugug) [inline CONTRIBUTING.md]: And yah, `npm test`

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -72,71 +92,79 @@ By making a contribution to this project, I certify that:
...
-* Include relevant Emoji from our [Emoji cheatsheet](#emoji-cheatsheet)
+* Limit the first line (not including emoji) to 72 characters or less
+
+Example commits may look something like the following:
+
+`git commit -m ":art: Refactor lookup system"`
+
+`git commit -m ":lock: Fix XSS vulnerability"`
+
+`git commit -m ":new::boom: Require author in content types" -m "Previously working content types will now throw if author is not present, so this is a breaking change"`
```
Comment (scottnath) [inline CONTRIBUTING.md]: for instance...did not know we could use two emoji...

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -72,71 +92,79 @@ By making a contribution to this project, I certify that:
...
 ### Branching Model
 
 * Branches must be made off of the most current `master` branch of a repository
 * Branch names should be descriptive, describing what is being done in that branch
 * Pull requests must be made back in to the `master` branch for a repository
 * The following branch prefixes should be used when creating a new branch:
-  * `hotfix/` - bug fixes that got through and need to be squashed
-  * `release/` - for releases
-  * `feature/` - update to or new functionality
+  * `hotfix/` - [bug fixes](#bug-repott) for bugs got through and need to be squashed
```
Comment (scottnath) [inline CONTRIBUTING.md]: `bug-report`

File: CONTRIBUTING.md (hunk 3 above)
Comment (Snugug) [inline CONTRIBUTING.md]: I _think_ that wiki has been superseded by these contributing guidelines? If not, we probably should supersede them with this so it's not buried in our wiki

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -72,71 +92,79 @@ By making a contribution to this project, I certify that:
...
+  * `feature/` - update to or [new functionality](#functionality-requests)
```
Comment (scottnath) [inline CONTRIBUTING.md]: * `feature/` - update to existing code or [new functionality](#functionality-requests)

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -72,71 +92,79 @@ By making a contribution to this project, I certify that:
...
 
-## Creating a New Version
+### Time to Merge
 
-Versioning is done through [SEMVER](http://semver.org/). When issuing a new version, create a [release branch](#branching-model) with the version's name in to the `master` branch. The release branch should bump the version number accordingly and [update the changelog](#maintaining-the-changelog). Once merged, create a new tag with `v` prefixed with the version's name from that branch.
+In order for a pull request to be merged, it must meet the following criteria:
 
-For instance, when creating version `1.1.0`, start by creating a release branch `release/1.1.0` that includes the version bump and changelog updates, merge it in to `master` and create a tag `v1.1.0` from branch `release/1.1.0`.
+* The author(s) of the code have signed off on the [Developer Certificate of Origin](#developer-certificate-of-origin) for this pull request
```
Comment (scottnath) [inline CONTRIBUTING.md]: would prefer a numbered list for this.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -72,71 +92,79 @@ By making a contribution to this project, I certify that:
...
+Versioning is done through [SEMVER](http://semver.org/). Whenever new code is merged in to the `master` branch, a new semantic release will automatically be created on NPM and in the project's GitHub Releases. The GitHub Release will also include a changelog based off of the commit messages of that release. The release version and changelog are determined by the [punchcard-semantic-release](https://github.com/punchcard-cms/semantic-release) module, and released using the [semantic-release](https://github.com/semantic-release/semantic-release) module based off of the [emoji](#emoji-cheatsheet) used in the commit messages for each commit. The following emoji correspond to different semantic versions, all other emoji are considered non-semantic:
 
-Changelog messages do not need to cover each individual commit made, but rather should have individual summaries of the changes made. Changelog messages should be written in the same style as [commit messages](#git-commit-messages).
+**Major Version**
```
Comment (scottnath) [inline CONTRIBUTING.md]: HA!!! Here it is

File: CONTRIBUTING.md (hunk 6 above)
Comment (Snugug) [inline CONTRIBUTING.md]: Keep going, that's under releases. We can link to that from here, though

File: CONTRIBUTING.md (hunk 7 above)
Comment (Snugug) [inline CONTRIBUTING.md]: Yup, that's why I'm updating this.

File: CONTRIBUTING.md (hunk 10 above)
Comment (Snugug) [inline CONTRIBUTING.md]: OOC, why?

File: CONTRIBUTING.md (hunk 11 above)
Comment (scottnath) [inline CONTRIBUTING.md]: This only covers nine of the 12 emoji. Should the other missing ones be in `In-progress release` too?

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -72,71 +92,79 @@ By making a contribution to this project, I certify that:
...
+There are generally two kinds of issues: functionality requests and bugs. Each is treated slightly differently, and there is a template for each.
 
-**Title**
+#### Functionality Requests
 
-    {{persona}}: {{need}}
+[Request New Functionality](https://github.com/punchcard-cms/punchcard/issues/new)
 
-**Body**
+New functionality always exists in relation to a specific type of user. When requesting new functionality, include the user persona the functionality relates to, the need for that functionality, and the rationale as to why this is needed. Titles should be written with the user persona first, then the a short description of their need (5-10 words).
```
Comment (scottnath) [inline CONTRIBUTING.md]: functionality functionality functionality functionality [functionality](http://www.thesaurus.com/browse/functionality)

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -72,71 +92,79 @@ By making a contribution to this project, I certify that:
...
-      Scenario: {{scenario}}
-        Given {{thing}}
-        When {{action}}
-        Then {{result}}
-    ```
-    
-    ---
-    
-    - [ ] {{feature}}/{{scenario}}
+New functionality also needs their requirements for being considered complete written out. To do so, features and scenarios should be written to describe the expected behavior in [Gherkin](https://github.com/cucumber/cucumber/wiki/Gherkin) syntax. Each feature and scenario par should have a [task list](https://github.com/blog/1375%0A-task-lists-in-gfm-issues-pulls-comments) created to track their completion. The issue is considered **done** when all scenarios defined have been completed with passing tests written.
```
Comment (scottnath) [inline CONTRIBUTING.md]: spelling: `scenario pair should`
rewrite? `The issue is considered **done** only when _all_ scenarios defined have been completed, code is covered with testing, and all tests pass.`

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -205,7 +191,7 @@ When creating creating commits or updating the CHANGELOG, please **start** the c
 * :bug: `:bug:` when fixing a bug
 * :memo: `:memo:` when writing documentation
 * :shirt: `:shirt:` when removing linter warnings
-* :art: `:art:` when improving the format/structure of the code
+* :art: `:art:` when improving the format/structure of the code or visual style
```
Comment (scottnath) [inline CONTRIBUTING.md]: phrase? `visual styling`?

File: CONTRIBUTING.md (hunk 1 above)
Comment (scottnath) [inline CONTRIBUTING.md]: Then the first one, but like this?

`In order to expedite the process of reviewing and merging pull requests, please adhere to _all_ of the following guidelines.`

File: CONTRIBUTING.md (hunk 4 above)
Comment (scottnath) [inline CONTRIBUTING.md]: moar explicit is better I believe

File: CONTRIBUTING.md (hunk 5 above)
Comment (scottnath) [inline CONTRIBUTING.md]: Different example please. Something less debatable that would work without clarification.

File: CONTRIBUTING.md (hunk 3 above)
Comment (scottnath) [inline CONTRIBUTING.md]: agreed. I hope we can phase out the wiki in general for that can be maintained via PRs...old news is old!

File: CONTRIBUTING.md (hunk 6 above)
Comment (scottnath) [inline CONTRIBUTING.md]: Yup, saw them...add link here for emoji below

File: CONTRIBUTING.md (hunk 10 above)
Comment (scottnath) [inline CONTRIBUTING.md]: Might be just my preference. Numbers feel like checkboxes in my mind.

Comment (Snugug) [inline CONTRIBUTING.md]: Okay, gonna keep it then

File: CONTRIBUTING.md (hunk 11 above)
Comment (Snugug) [inline CONTRIBUTING.md]: The other 3 emoji are `:white_check_mark:` for adding tests, `:green_heart:` fixing the CI build, and `:shipit:` for rolling a new release. The last one no one should be hand-writing anymore (and is circular when used for a semantic release), and the first two don't contribute to a new release (and aren't in [analyze](https://github.com/punchcard-cms/semantic-release/blob/master/analyze.js)). If you'd like them added, bring it up there plz

**Comment by Snugug** [comment]
@scottnath look at all those updates!

**Pull request merged**